pkg/server/user_connections.go

```diff
@@ -0,0 +1,132 @@
+// Note: increaseUserConnectionsCount() is called when create connections only, and not in 'changeUser'(COM_CHANGE_USER).
+// In mysql, 'COM_CHANGE_USER' will only reset the session state (such as permissions, default database, etc.),
+// but will not change the ownership of the number of connections.
+func (cc *clientConn) increaseUserConnectionsCount() error {
+	var (
+		user       = cc.ctx.GetSessionVars().User
+		targetUser = user.String()
+		globaLimit = vardef.MaxUserConnectionsValue.Load()
+	)
+
+	// check the number of connections again.
+	pm := privilege.GetPrivilegeManager(cc.ctx.Session)
+	userLimit, err := pm.GetUserResources(user.AuthUsername, user.AuthHostname)
+	if err != nil {
+		return err
+	}
+
+	cc.server.userResLock.Lock()
+	defer cc.server.userResLock.Unlock()
+	ur, ok := cc.server.userResource[targetUser]
+	if !ok {
+		// if not exist in userResource,
+		userHost := &userResourceLimits{
+			connections: 1,
+		}
+		cc.server.userResource[targetUser] = userHost
+		return nil
+	}
+
+	// check the global variables `MAX_USER_CONNECTIONS` and the max_user_conections in mysql.user
+	// with current count of active connection.
+	if (userLimit > 0 && ur.connections >= int(userLimit)) ||
+		(globaLimit > 0 && ur.connections >= int(globaLimit)) {
+		return servererr.ErrTooManyUserConnections.GenWithStackByArgs(targetUser)
+	}
+
+	ur.connections++
+	return nil
+}
+
+// decreaseUserConnectionCount decreases the count of connections when user logout the database.
+func (cc *clientConn) decreaseUserConnectionCount() {
+	user := cc.ctx.GetSessionVars().User
+	targetUser := user.String()
+
+	cc.server.userResLock.Lock()
+	defer cc.server.userResLock.Unlock()
+	ur, ok := cc.server.userResource[targetUser]
+	if ok && ur.connections > 0 {
+		ur.connections--
+		if ur.connections <= 0 {
+			delete(cc.server.userResource, targetUser)
+		}
+	}
+}
+
+// getUserConnectionCount gets the count of connections.
+func (cc *clientConn) getUserConnectionCount(user *auth.UserIdentity) int {
+	targetUser := user.String()
+	cc.server.userResLock.Lock()
+	defer cc.server.userResLock.Unlock()
+
+	ur, ok := cc.server.userResource[targetUser]
+	if ok {
+		return ur.connections
+	}
+
+	return 0
+}
+
+// checkUserConnectionCount checks whether the count of connections exceeds the limits.
+func (cc *clientConn) checkUserConnectionCount(host string) error {
+	authUser, err := cc.ctx.MatchIdentity(context.Background(), cc.user, host)
+	if err != nil {
+		return err
+	}
+
+	pm := privilege.GetPrivilegeManager(cc.ctx.Session)
+	connections, err := pm.GetUserResources(authUser.Username, authUser.Hostname)
+	if err != nil {
+		return err
+	}
+
+	if connections == 0 && vardef.MaxUserConnectionsValue.Load() == 0 {
+		return nil
+	}
+
+	conns := int64(cc.getUserConnectionCount(authUser))
+	if (connections > 0 && conns >= connections) || (connections == 0 && conns >= int64(vardef.MaxUserConnectionsValue.Load())) {
```

Comment:
If `@@max_user_connection` variable > the user's max_user_connection attribute, what's the behavior?

For example, user attribute specify the value to 10, @@max_user_connection is set to 5,
the 10 connections are allowed?

Reply:
If @@max_user_connection > 0(is limited) and max_user_connection attribute > 0 (is limited). The max_user_connection attribute has a higher priority than @@max_user_connection, and  ignore theirs numbers